homeassistant/components/alert.py

```diff
@@ -0,0 +1,283 @@
+
+
+@callback
+def async_turn_on(hass, entity_id):
+    """Async reset the alert."""
+    data = {ATTR_ENTITY_ID: entity_id}
+    hass.async_add_job(
+        hass.services.async_call(DOMAIN, SERVICE_TURN_ON, data))
+
+
+def turn_off(hass, entity_id):
+    """Acknowledge alert."""
+    run_callback_threadsafe(hass.loop, async_turn_off, hass, entity_id)
+
+
+@callback
+def async_turn_off(hass, entity_id):
+    """Async acknowledge the alert."""
+    data = {ATTR_ENTITY_ID: entity_id}
+    hass.async_add_job(
+        hass.services.async_call(DOMAIN, SERVICE_TURN_OFF, data))
+
+
+def toggle(hass, entity_id):
+    """Toggle acknowledgement of alert."""
+    run_callback_threadsafe(hass.loop, async_toggle, hass, entity_id)
+
+
+@callback
+def async_toggle(hass, entity_id):
+    """Async toggle acknowledgement of alert."""
+    data = {ATTR_ENTITY_ID: entity_id}
+    hass.async_add_job(
+        hass.services.async_call(DOMAIN, SERVICE_TOGGLE, data))
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Setup alert component."""
+    alerts = config.get(DOMAIN)
+    all_alerts = {}
+
+    @asyncio.coroutine
+    def async_handle_alert_service(service_call):
+        """Handle calls to alert services."""
+        alert_ids = service.extract_entity_ids(hass, service_call)
+
+        for alert_id in alert_ids:
+            alert = all_alerts[alert_id]
+            if service_call.service == SERVICE_TURN_ON:
+                yield from alert.async_turn_on()
+            elif service_call.service == SERVICE_TOGGLE:
+                yield from alert.async_toggle()
+            else:
+                yield from alert.async_turn_off()
+
+    # setup alerts
+    for entity_id, alert in alerts.items():
+        entity = Alert(hass, entity_id,
+                       alert[CONF_NAME], alert[CONF_ENTITY_ID],
+                       alert[CONF_STATE], alert[CONF_REPEAT],
+                       alert[CONF_REPEAT_MIN], alert[CONF_REPEAT_MAX],
+                       alert[CONF_SKIP_FIRST], alert[CONF_NOTIFIERS],
+                       alert[CONF_CAN_ACK], alert[CONF_BACKOFF])
+        all_alerts[entity.entity_id] = entity
+
+    # read descriptions
+    descriptions = yield from hass.loop.run_in_executor(
+        None, load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml'))
+    descriptions = descriptions.get(DOMAIN, {})
+
+    # setup service calls
+    hass.services.async_register(
+        DOMAIN, SERVICE_TURN_OFF, async_handle_alert_service,
+        descriptions.get(SERVICE_TURN_OFF), schema=ALERT_SERVICE_SCHEMA)
+    hass.services.async_register(
+        DOMAIN, SERVICE_TURN_ON, async_handle_alert_service,
+        descriptions.get(SERVICE_TURN_ON), schema=ALERT_SERVICE_SCHEMA)
+    hass.services.async_register(
+        DOMAIN, SERVICE_TOGGLE, async_handle_alert_service,
+        descriptions.get(SERVICE_TOGGLE), schema=ALERT_SERVICE_SCHEMA)
+
+    for alert in all_alerts.values():
```

Comment:
I'm not an expert with asyncio, but, did you intend to be calling the async_update_ha_state method? Did you intend for that to be more like this:

```python
tasks = [alert.async_update_ha_state for alert in all_alerts.values()]
if tasks:
    yield from asyncio.wait(tasks, loop=hass.loop)
```

Otherwise, I think you are just running the update_ha_state methods synchronously. If my presumption is correct (although it may not be), that doesn't work because the Alert class is not hashable. I think, what I have should be relatively asynchronous. However, I am not very well versed with asyncio yet.